fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -2182,46 +2185,80 @@ Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext searchCont
             queryParameters.addAll(logicalIdReferenceQueryParameters);
             Map<String, String> logicalIdToTypeMap = new HashMap<>();
 
+            FHIRPathEvaluator evaluator = FHIRPathEvaluator.evaluator();
+
+
             // Loop through the resources, looking for versioned references and references to multiple resource types for the same logical ID
             for (Resource resource : matchResources) {
-                FHIRPathEvaluator evaluator = FHIRPathEvaluator.evaluator();
-                EvaluationContext evaluationContext = new EvaluationContext(resource);
-                for (QueryParameter queryParameter : queryParameters) {
-                    SearchParameter searchParameter = searchParameterMap.get(queryParameter);
-                    if (searchParameter == null) {
-                        searchParameter = SearchUtil.getSearchParameter(resource.getClass(), queryParameter.getCode());
+                // A flag that indicates whether we need to take a closer look at the reference values or not
+                boolean needsEval = false;
+
+                // If any of the reference values are "versioned"
+                // then we'll need to check that they aren't used for chaining
+                // TODO Should we pass the previously-gathered set of references into the method instead?
+                CollectingVisitor<Reference> refCollector = new CollectingVisitor<>(Reference.class);
+                resource.accept(refCollector);
+                List<Reference> references = refCollector.getResult();
+                for (Reference ref : references) {
```

Comment:
You can end this loop as soon as needsEval is true. It's a shame that this iteration stuff doesn't have a boolean condition option - that would be a nice addition to the language.

Reply:
easy enough just to break, will add quick

Reply:
done